src/ops/arithmetic.rs

```diff
@@ -0,0 +1,875 @@
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1.0, 2.0, 3.0], &[3]);
+    /// let mut b = a.log1p_device(Default::default());
+    ///
+    /// b.eval();
+    /// let b_data: &[f32] = b.as_slice();
+    /// // b_data == [0.6931472, 1.0986123, 1.3862944]
+    /// ```
+    #[default_device]
+    pub fn log1p_device(&self, stream: StreamOrDevice) -> Array {
+        unsafe { Array::from_ptr(mlx_sys::mlx_log1p(self.c_array, stream.as_ptr())) }
+    }
+
+    /// Matrix multiplication.
+    ///
+    /// Perform the (possibly batched) matrix multiplication of two arrays. This function supports
+    /// broadcasting for arrays with more than two dimensions.
+    ///
+    /// - If the first array is 1-D then a 1 is prepended to its shape to make it
+    ///   a matrix. Similarly, if the second array is 1-D then a 1 is appended to its
+    ///   shape to make it a matrix. In either case the singleton dimension is removed
+    ///   from the result.
+    /// - A batched matrix multiplication is performed if the arrays have more than
+    ///   2 dimensions.  The matrix dimensions for the matrix product are the last
+    ///   two dimensions of each input.
+    /// - All but the last two dimensions of each input are broadcast with one another using
+    ///   standard <doc:broadcasting>.
+    ///
+    /// # Example
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1, 2, 3, 4], &[2, 2]);
+    /// let b = Array::from_slice(&[-5.0, 37.5, 4., 7., 1., 0.], &[2, 3]);
+    ///
+    /// // produces a [2, 3] result
+    /// let mut c = a.matmul_device(&b, Default::default());
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to multiply
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn matmul_device(&self, other: &Array, stream: StreamOrDevice) -> Array {
+        unsafe {
+            Array::from_ptr(mlx_sys::mlx_matmul(
+                self.c_array,
+                other.c_array,
+                stream.as_ptr(),
+            ))
+        }
+    }
+
+    /// Element-wise reciprocal.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1.0, 2.0, 4.0], &[3]);
+    /// let mut b = a.reciprocal_device(Default::default());
+    ///
+    /// b.eval();
+    /// let b_data: &[f32] = b.as_slice();
+    /// // b_data == [1.0, 0.5, 0.25]
+    /// ```
+    #[default_device]
+    pub fn reciprocal_device(&self, stream: StreamOrDevice) -> Array {
+        unsafe { Array::from_ptr(mlx_sys::mlx_reciprocal(self.c_array, stream.as_ptr())) }
+    }
+
+    /// Round to the given number of decimals.
+    ///
+    /// # Params
+    ///
+    /// - decimals: number of decimals to round to - default is 0 if not provided
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn round_device(&self, decimals: impl Into<Option<i32>>, stream: StreamOrDevice) -> Array {
```

Comment:
What happens on the c++ side if `decimals < 0`? Does it throw an exception?

Reply:
I think this is fine as long as the c++ side doesn't just throw and exception. Maybe we should test that

Reply:
turns out this is valid in mlx, it just returns an array of zeros